core/src/com/unciv/ui/components/widgets/UncivTextField.kt

```diff
@@ -0,0 +1,192 @@
+ *
+ * - On Android, manages on-screen keyboard visibility and reacts to how the on-screen keyboard reduces screen space.
+ * - Tries to scroll the field into view when receiving focus using an ascendant ScrollPane.
+ * - If view of the field is still obscured, show am "emergency" Popup instead (can help when on Android the on-screen keyboard is large and the screen small).
+ * - If this TextField handles the Tab key (see [keyShortcuts]), its [focus navigation feature][TextField.next] is disabled (otherwise it would search for another TextField in the same parent to give focus to, and remove the on-screen keyboard if it finds none).
+ * - All parameters are supported on all platforms.
+ *
+ * @param hint The text that should be displayed in the text field when no text is entered and it does not have focus, will automatically be translated. Also shown as Label in the "emergency" Popup.
+ * @param preEnteredText the text initially entered within this text field.
+ * @param onFocusChange a callback that will be notified when this TextField loses or gains the [keyboardFocus][com.badlogic.gdx.scenes.scene2d.Stage.keyboardFocus].
+ *        Receiver is the field, so you can simply use its elements. Parameter `it` is a Boolean indicating focus was received.
+ */
+class UncivTextField(
+    hint: String,
+    preEnteredText: String = "",
+    private val onFocusChange: (TextField.(Boolean) -> Unit)? = null
+) : TextFieldWithFixes(preEnteredText, BaseScreen.skin) {
+    private val isAndroid = Gdx.app.type == Application.ApplicationType.Android
+    private val hideKeyboard = { Gdx.input.setOnscreenKeyboardVisible(false) }
+
+    init {
+        messageText = hint.tr()
+
+        addListener(UncivTextFieldFocusListener())
+
+        if (isAndroid) addListener(VisibleAreaChangedListener())
+    }
+
+    private inner class UncivTextFieldFocusListener : FocusListener() {
+        override fun keyboardFocusChanged(event: FocusEvent, actor: Actor, focused: Boolean) {
+            if (focused) {
+                scrollAscendantToTextField()
+                if (isAndroid) {
+                    addPopupCloseListener()
+                    Gdx.input.setOnscreenKeyboardVisible(true)
+                }
+            }
+            onFocusChange?.invoke(this@UncivTextField, focused)
+        }
+    }
+
+    private inner class VisibleAreaChangedListener : InputListener() {
+        private val events = EventBus.EventReceiver()
+        init {
+            events.receive(UncivStage.VisibleAreaChanged::class) {
+                if (stage == null || !hasKeyboardFocus()) return@receive
+                Concurrency.run {
+                    // If anything resizes, it also does so with this event. So we need to wait for that to finish to update the scroll position.
+                    delay(100)
+                    withGLContext {
+                        if (stage == null) return@withGLContext
```

Comment:
In what situation would the stage have previously not been null, but would now be null? If the actor was removed?